File: app/src/main/java/io/weave/client/core/diagnostics/DiagnosticSafeSummary.kt
package io.weave.client.core.diagnostics

import io.weave.client.core.ipquality.IpQualityReport

/** A structural allowlist, not regex redaction of secrets or free-form error messages. */
object DiagnosticSafeSummary {
    fun build(
        version: String,
        local: PrivacyObservationReport,
        ip: IpQualityReport?,
        sites: CommonEndpointReport?,
        historical: Boolean,
        browserMeasured: Boolean,
        browserCandidates: Int,
    ): String = buildString {
        appendLine("Weave Android · diagnostic summary v1")
        appendLine("app=" + (version.takeIf { it.matches(Regex("[A-Za-z0-9.+-]{1,48}")) } ?: "unknown"))
        appendLine("historical=$historical")
        appendLine("local_verified=${local.verifiedCount}; local_attention=${local.attentionCount}")
        appendLine("ip_measured=${ip != null}; ipv4_present=${!ip?.ipv4.isNullOrBlank()}; ipv6_present=${!ip?.ipv6.isNullOrBlank()}")
        appendLine("browser_measured=$browserMeasured; ice_candidates=${browserCandidates.coerceIn(0, 999)}")
        CommonEndpointProbe.COMMON_ENDPOINTS.forEach { trusted ->
            val row = sites?.results?.firstOrNull { it.endpoint.id == trusted.id } ?: return@forEach
            append(trusted.id).append(": ").append(row.state.name)
            append("; http=").append(row.statusCode?.takeIf { it in 100..599 } ?: "unknown")
            append("; ms=").append(row.latencyMs?.takeIf { it in 1..60_000 } ?: "unknown")
            append("; failure=").append(row.failure?.name ?: "none").appendLine()
        }
        append("Snapshot only. No IP addresses, node/subscription names, credentials, fingerprints or raw errors. Not proof of leak-free networking.")
    }
}

File: windows/src/Weave.Windows.Core/DiagnosticSafeSummary.cs
using System.Text;

namespace Weave.Windows.Core;

/// <summary>Only typed, allowlisted evidence is exported. Never copy raw Result or error text.</summary>
public static class DiagnosticSafeSummary
{
    public static string Build(IEnumerable<ProbeResult> evidence, bool historical)
    {
        var rows = evidence.Take(64).ToArray();
        var text = new StringBuilder("Weave Windows · diagnostic summary v1\n")
            .Append("historical=").Append(historical).AppendLine();
        foreach (var (name, ipv6) in new[] { ("代理出口 IPv4", false), ("代理出口 IPv6", true) })
        {
            var row = rows.FirstOrDefault(item => item.Name == name);
            text.Append(ipv6 ? "ipv6" : "ipv4").Append("_measured=").Append(row is not null).Append("; present=")
                .Append(row?.ExitVerified == true).Append("; failure=").Append(row?.Failure?.ToString() ?? "none").AppendLine();
        }
        foreach (var target in NetworkDiagnostics.Targets)
        {
            var row = rows.FirstOrDefault(item => item.Name == target.Name);
            if (row is null) continue;
            text.Append(target.Name).Append(": ").Append(row.EndpointVerified ? "VERIFIED" : "UNCONFIRMED")
                .Append("; http=").Append(row.HttpStatus is >= 100 and <= 599 ? row.HttpStatus.Value.ToString() : "unknown")
                .Append("; ms=").Append(row.Milliseconds is >= 0 and <= 60_000 ? row.Milliseconds.Value.ToString() : "unknown")
                .Append("; failure=").Append(row.Failure?.ToString() ?? "none").AppendLine();
        }
        return text.Append("Snapshot only. No IP addresses, node/subscription names, credentials, fingerprints or raw errors. Not proof of leak-free networking.").ToString();
    }
}
